proposals/0000-arraybuffer-support-in-turbomodules.md

```diff
@@ -0,0 +1,381 @@
+
+However, Codegen provides more functionality and broader support for types that can be shared across platforms. Below you can find the list (probably not full) of them:
+
+- Promises
+
+```ts
+export interface Spec extends TurboModule {
+  getBuffer(): ArrayBuffer;
+  getAsyncBuffer(): Promise<ArrayBuffer>;
+}
+```
+
+- Structs
+
+```ts
+export type ObjectStruct = {
+  a: number;
+  b: string;
+  c?: ?string;
+};
+export type BufferStruct = {
+  text: string;
+  value: ArrayBuffer;
+};
+```
+
+- Constants (special case of Structs)
+- Unions (can use Struct underneath)
+
+```ts
+export type UnionFloat = 1.44 | 2.88 | 5.76;
+export type UnionObject = { value: number } | { buffer: ArrayBuffer };
+```
+
+- Events
+
+```ts
+export interface Spec extends TurboModule {
+  onEvent: CodegenTypes.EventEmitter<string>;
+  onData: CodegenTypes.EventEmitter<ArrayBuffer>;
+}
+```
+
+As illustrated, the `ArrayBuffer` type can be used in many contexts; a simple input/output parameter is only one of them. Handling each separately will be required to provide full and consistent support across all platforms. In order to pinpoint potential problems we investigated the topic at this stage. Here is a breakdown of the results by platform.
+
+#### C++
+
+It looks like the current implementation of `AsyncEventEmitter` requires an argument type to be copyable. That is not the case for `jsi::ArrayBuffer`. This will probably require deeper investigation and perhaps more complex changes to the bridging part of the `ReactCommon` library.
+
+#### Java
+
+While in C++ and Objective-C data can be easily shared between JS and Native, Java stores the data as a `folly::dynamic` map on the Native side. The `folly` library has support for data buffers (class `IOBuf`). This means that JS buffers can be stored on the Native side, but its implementation will be more challenging. Moreover, classes responsible for storing variables, such as `NativeMap` or `NatviveArray`, have a rich inheritance tree and are widely used across the JNI files. Adding storage for buffers to them will require changes to a large number of `ReactAndroid` JNI and Java/Kotlin files. These changes are required to add support for e.g. Promises or Structs.
+
+#### Objective-C
+
+No obstacles were found for this platform.
+
+#### Conclusion
+
+To conclude, supporting only simple function input/output parameters is straightforward, but extending that support to cover all Codegen functionalities across every platform is significantly more complex. However, integrating Promises and asynchronous operations may be crucial for supporting binary-heavy use cases. This leads to the question:
+
+> [!IMPORTANT]
+> Should this RFC focus on introducing only the basic and most valuable synchronous support for ArrayBuffer, or should it aim for full coverage of all possible Codegen use cases, including asynchronous operations, despite the higher complexity and the impact on more files (especially on Android)?
```

Comment:
In my opinion, the RFC should aim to cover all use cases and describe the complete implementation plan. We can then approach it incrementally, breaking the work into smaller, manageable PRs. It might also make sense to update the Adoption Strategy section with a detailed roll-out plan that outlines the milestones we’ll follow once this RFC is approved.